### ISSUE TITLE
Dynamically injecting data sources into form fields?

### ISSUE THREAD

**Issue by tnorthcutt** [formio/formio.js#2464]
As part of our Form.io integration, we're going to have a number of complex rules involved in determining e.g. which options are available in select fields (that's just one example). It seems like this complexity will necessitate dynamically injecting data sources into some of our form fields, in response to input in earlier form fields.

Is this a route that is feasible, and/or recommended? I'm wondering if [custom components](https://github.com/formio/formio.js/wiki/Custom-Components-API) is the way to go, or if there's another approach that's possible or makes more sense.

Thanks for any guidance you're able to provide.

**Comment by randallknutson** [comment]
This does depend on exactly how you want this to work. We do have several built in features that should help you get most use cases taken care of. First, there is "calculated value" where you can use javascript to set the value of a field based on previous fields. In addition, we have options in the select component that will change the options of the select based on other fields. You can do this with either javascript (using the Custom option) or an API query (using the URL option). In addition, you can change the options in Radio/Selectboxes components by using field logic and modifying the component definition as the action.

Basically, there is a lot already built in to support what you are asking for but at the end of the day, if it is really complex you may need to use custom components.

**Comment by bh-adrienne** [comment]
@tnorthcutt  I don't know which platform you're working in, but I've been using the Angular Formio renderer and have had no issues with getting the Form json, modifying it, and then reloading/refreshing the form with events or logic in my Angular components since the Form.io model really is flat JSON. I haven't had to add any custom Form.io components, just selected the fields I wanted from the model, updated them, and used Form.io's refresh events.

**Comment by raymondKelly** [comment]
@bh-adrienne I think one concern I have with that approach is that having to perform a full rerender when only one component changes might be a little expensive on a large form but perhaps that is unavoidable as the forms internal logic would need to run if a component is changed. A function such as updateComponent or updateForm where you could push in some changes to the form and only that component would be rerendered and only the associated internal rules would be rerun.

**Comment by raymondKelly** [comment]
@tnorthcutt I am currently doing a POC with this form renderer and I think I have a similar use case to you. Did you end up finding a solution that satisfied your needs?
